File: src/main/java/com/github/hirsivaja/ip/ipv6/Ipv6Payload.java
package com.github.hirsivaja.ip.ipv6;

import com.github.hirsivaja.ip.*;
import com.github.hirsivaja.ip.icmpv6.Icmpv6Payload;
import com.github.hirsivaja.ip.tcp.TcpSegment;
import com.github.hirsivaja.ip.udp.UdpDatagram;

import java.nio.ByteBuffer;

public sealed interface Ipv6Payload extends IpPayload permits Icmpv6Payload, EncapsulationPayload, Ipv6Payload.Generic, TcpSegment, UdpDatagram {

    static Ipv6Payload decode(ByteBuffer in, boolean ensureChecksum, Ipv6Header header) {
        return switch (header.lastNextHeader()) {
            case IpProtocols.TCP -> TcpSegment.decode(in, ensureChecksum, header);
            case IpProtocols.UDP -> UdpDatagram.decode(in, ensureChecksum, header);
            case IpProtocols.ICMPV6 -> Icmpv6Payload.decode(in, ensureChecksum, header);
            case IpProtocols.IPV6_ENCAPSULATION -> EncapsulationPayload.decode(in, ensureChecksum);
            default -> Ipv6Payload.Generic.decode(in, header.lastNextHeader());
        };
    }

    record Generic(IpProtocol protocol, ByteArray data) implements Ipv6Payload {

        @Override
        public void encode(ByteBuffer out) {
            out.put(data.array());
        }

        @Override
        public int length() {
            return data.array().length;
        }

        public static Ipv6Payload decode(ByteBuffer in, IpProtocol protocol) {
            byte[] data = new byte[in.remaining()];
            in.get(data);
            return new Generic(protocol, new ByteArray(data));
        }
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/payload/Ipv6Payload.kt
package com.github.hirsivaja.ip.packet.payload

import com.github.hirsivaja.ip.DataArray
import com.github.hirsivaja.ip.packet.header.Ipv6Header
import com.github.hirsivaja.ip.packet.payload.tcp.TcpSegment
import com.github.hirsivaja.ip.packet.payload.udp.UdpDatagram
import com.github.hirsivaja.ip.protocol.IpProtocol
import com.github.hirsivaja.ip.protocol.IpProtocols
import java.nio.ByteBuffer

interface Ipv6Payload : IpPayload {
    companion object {
        fun decode(buf: ByteBuffer, ensureChecksum: Boolean, header: Ipv6Header) : Ipv6Payload {
            return when (header.protocol()) {
                IpProtocols.TCP -> TcpSegment.decode(buf, ensureChecksum, header)
                IpProtocols.UDP -> UdpDatagram.decode(buf, ensureChecksum, header)
                else -> Generic.decode(buf, header.protocol())
            } as Ipv6Payload
        }
    }

    data class Generic(val protocol: IpProtocol, val data: DataArray) : Ipv6Payload {
        constructor(protocol: IpProtocol, data: ByteArray) : this(protocol, DataArray(data))

        override fun encode(out: ByteBuffer) {
            out.put(data.data)
        }

        override fun length(): Int {
            return data.length()
        }

        companion object {
            fun decode(buf: ByteBuffer, protocol: IpProtocol): IpPayload {
                val data = ByteArray(buf.remaining())
                buf[data]
                return Generic(protocol, data)
            }
        }
    }
}
